### PR TITLE
Fix #245 File header rule: index out of range [0] with length 0

### LINKED ISSUE #245: File Header rule panic with "index out of range"
**Describe the bug**
I got some files without comments and when 'revive' apply 'file-header' rule.

**To Reproduce**
1. I updated revive `go get -u github.com/mgechev/revive`
2. I run it with the following flags & configuration file:

```shell
revive --config ./revive.toml --formatter friendly --exclude ./vendor/... ./... 
```

```toml
ignoreGeneratedHeader = true
severity = "warning"
confidence = 0.8
errorCode = 1
warningCode = 2

[rule.file-header]
  arguments =["Copyright 1999-2019. All rights reserved."]
```

**Expected behavior**
We should not get panic.

**Logs**
```
panic: runtime error: index out of range [0] with length 0

goroutine 54 [running]:
github.com/mgechev/revive/rule.lintFileHeader.Visit(0xc0002d37c0, 0xc00016ce80, 0xc0004840a0, 0xc0003523c0, 0x13c39e0, 0xc00016ce80, 0x8, 0xc000444230)
	/Users/dmitriysemin/go/src/github.com/mgechev/revive/rule/file-header.go:55 +0x436
go/ast.Walk(0x13c0d60, 0xc0003ae320, 0x13c39e0, 0xc00016ce80)
	/usr/local/Cellar/go/1.13.4/libexec/src/go/ast/walk.go:52 +0x66
github.com/mgechev/revive/rule.(*FileHeaderRule).Apply(0x15d5f48, 0xc0002d37c0, 0xc000076f00, 0x1, 0x1, 0xc000335b60, 0xc0002a5980, 0x130e680)
	/Users/dmitriysemin/go/src/github.com/mgechev/revive/rule/file-header.go:37 +0x161
github.com/mgechev/revive/lint.(*File).lint(0xc0002d37c0, 0xc000437390, 0x1, 0x1, 0x1, 0x3fe999999999999a, 0xc00009c990, 0x7, 0xc000079050, 0x1, ...)
	/Users/dmitriysemin/go/src/github.com/mgechev/revive/lint/file.go:108 +0x3b6
github.com/mgechev/revive/lint.(*Package).lint.func1(0xc000437390, 0x1, 0x1, 0x1, 0x3fe999999999999a, 0xc00009c990, 0x7, 0xc000079050, 0x1, 0x2, ...)
	/Users/dmitriysemin/go/src/github.com/mgechev/revive/lint/package.go:173 +0xa8
created by github.com/mgechev/revive/lint.(*Package).lint
	/Users/dmitriysemin/go/src/github.com/mgechev/revive/lint/package.go:172 +0x179

```

**Desktop (please complete the following information):**
 - OS: macOS Mojave 10.14.6
 - Version of Go: 1.13.4

### PR DESCRIPTION
**Pull request by freedemster**
Close #245

### CODE REVIEW

File: test/file-header_test.go [hunk 1]
```diff
@@ -27,4 +27,20 @@ func TestLintFileHeader(t *testing.T) {
...
+	testRule(t, "lint-file-header6", &rule.FileHeaderRule{}, &lint.RuleConfig{
+		Arguments: []interface{}{"foobar"},
+	})
+}
+
+func BenchmarkLintFileHeader(b *testing.B) {
+	benchRule(b, "lint-file-header1", &rule.FileHeaderRule{}, &lint.RuleConfig{
+		Arguments: []interface{}{"foobar"},
+	})
+	//var t *testing.T
```
Comment (chavacava) [inline test/file-header_test.go]: please avoid commented-out code in commits

File: test/import-blacklist_test.go [hunk 2]
```diff
@@ -17,10 +17,7 @@ func TestImportsBlacklist(t *testing.T) {
 
 func BenchmarkImportsBlacklist(b *testing.B) {
 	args := []interface{}{"crypto/md5", "crypto/sha1"}
-	var t *testing.T
-	for i := 0; i <= b.N; i++ {
-		testRule(t, "imports-blacklist", &rule.ImportsBlacklistRule{}, &lint.RuleConfig{
-			Arguments: args,
-		})
-	}
+	benchRule(b, "imports-blacklist", &rule.ImportsBlacklistRule{}, &lint.RuleConfig{
```
Comment (chavacava) [inline test/import-blacklist_test.go]: this code is not related with the fix for issue #246, please consider openning other PR

File: test/unused-param_test.go [hunk 3]
```diff
@@ -11,8 +11,5 @@ func TestUnusedParam(t *testing.T) {
 }
 
 func BenchmarkUnusedParam(b *testing.B) {
-	var t *testing.T
-	for i := 0; i <= b.N; i++ {
-		testRule(t, "unused-param", &rule.UnusedParamRule{})
-	}
+	benchRule(b, "unused-param", &rule.UnusedParamRule{})
```
Comment (chavacava) [inline test/unused-param_test.go]: this code is not related with the fix for issue #246, please consider openning another PR

File: test/utils.go [hunk 4]
```diff
@@ -19,14 +18,42 @@ import (
 	"github.com/pkg/errors"
 )
 
+const fixtureBasePath = "../fixtures/"
```
Comment (chavacava) [inline test/utils.go]: this code is not related with the fix for issue #246, please consider openning another PR

File: .gitignore [hunk 5]
```diff
@@ -1,3 +1,4 @@
+.idea
```
Comment (chavacava) [inline .gitignore]: please leave .gitignore as is

File: rule/file-header.go [hunk 6]
```diff
@@ -51,25 +51,34 @@ type lintFileHeader struct {
...
+)
+
 func (w lintFileHeader) Visit(_ ast.Node) ast.Visitor {
-	g := w.fileAst.Comments[0]
 	failure := lint.Failure{
 		Node:       w.fileAst,
 		Confidence: 1,
 		Failure:    "the file doesn't have an appropriate header",
 	}
+
```
Comment (chavacava) [inline rule/file-header.go]: We can take advantage of this PR to remove the unnecessary call to Visit. All the logic of this rule can be implemented without visiting the file AST. That is, it is possible to access and check the comments of the file at the `Apply` method.

File: test/file-header_test.go (hunk 1 above)
Comment (freedemster) [reply test/file-header_test.go]: Done

File: test/import-blacklist_test.go (hunk 2 above)
Comment (freedemster) [reply test/import-blacklist_test.go]: Done

File: test/unused-param_test.go (hunk 3 above)
Comment (freedemster) [reply test/unused-param_test.go]: Done

File: test/utils.go (hunk 4 above)
Comment (freedemster) [reply test/utils.go]: Done

File: .gitignore (hunk 5 above)
Comment (freedemster) [reply .gitignore]: Done

File: rule/file-header.go (hunk 6 above)
Comment (freedemster) [reply rule/file-header.go]: Done

**Comment by chavacava** [comment]
Hi @freedemster, thanks for the fix. Its looks OK. 
I´ll try to see why the CI is not passing and then merge your fix.

**Comment by freedemster** [comment]
Thank you too!

File: rule/file-header.go [hunk 7]
```diff
@@ -10,73 +9,57 @@ import (
...
+var (
+	multiRegexp  = regexp.MustCompile("^/\\*")
+	singleRegexp = regexp.MustCompile("^//")
+)
+
 // Apply applies the rule to given file.
 func (r *FileHeaderRule) Apply(file *lint.File, arguments lint.Arguments) []lint.Failure {
-	var failures []lint.Failure
-
 	header, ok := arguments[0].(string)
```
Comment (chavacava) [inline rule/file-header.go]: This line panics (index out of range) if the `arguments` slice is empty (i.e. when the rule is not configured appropriately)  
We need to check the length of the slice before accessing it. Something like
```go
if len(arguments) != 1 {
	panic(`invalid configuration for "file-header" rule`)
}
```

Comment (freedemster) [reply rule/file-header.go]: Done

**Pull request merged**